crates/oxc_linter/src/rules/eslint/no_unused_vars/symbol.rs

```diff
@@ -0,0 +1,225 @@
+    }
+
+    pub fn iter_relevant_parents(
+        &self,
+        node_id: AstNodeId,
+    ) -> impl Iterator<Item = &AstNode<'a>> + Clone + '_ {
+        self.nodes().iter_parents(node_id).skip(1).filter(|n| Self::is_relevant_kind(n.kind()))
+    }
+
+    pub fn iter_relevant_parent_and_grandparent_kinds(
+        &self,
+        node_id: AstNodeId,
+    ) -> impl Iterator<Item = (/* parent */ AstKind<'a>, /* grandparent */ AstKind<'a>)> + Clone + '_
+    {
+        let parents_iter = self
+            .nodes()
+            .iter_parents(node_id)
+            .map(AstNode::kind)
+            // no skip
+            .filter(|kind| Self::is_relevant_kind(*kind));
+
+        let grandparents_iter = parents_iter.clone().skip(1);
+
+        parents_iter.zip(grandparents_iter)
+    }
+
+    #[inline]
+    const fn is_relevant_kind(kind: AstKind<'a>) -> bool {
+        !matches!(kind, AstKind::ParenthesizedExpression(_))
+    }
+}
+
+impl<'s, 'a> Symbol<'s, 'a> {
+    /// Is this [`Symbol`] exported?
+    ///
+    /// NOTE: does not support CJS right now.
+    pub fn is_exported(&self) -> bool {
+        let is_in_exportable_scope = self.is_root() || self.is_in_ts_namespace();
+        (is_in_exportable_scope
+            && (self.flags.contains(SymbolFlags::Export)
+                || self.semantic.module_record().exported_bindings.contains_key(self.name())))
+            || self.in_export_node()
+    }
+
+    #[inline]
+    fn is_in_ts_namespace(&self) -> bool {
+        self.scopes().get_flags(self.scope_id()).is_ts_module_block()
+    }
+
+    /// We need to do this due to limitations of [`Semantic`].
+    fn in_export_node(&self) -> bool {
+        for parent in self.nodes().iter_parents(self.declaration_id()).skip(1) {
+            match parent.kind() {
+                AstKind::ModuleDeclaration(module) => {
+                    return module.is_export();
+                }
+                AstKind::VariableDeclaration(_) => {
+                    continue;
+                }
+                _ => {
+                    return false;
+                }
+            }
+        }
+        false
+    }
+
+    #[inline]
+    pub fn is_in_jsx(&self) -> bool {
+        self.semantic.source_type().is_jsx()
+    }
+
+    #[inline]
+    pub fn is_in_ts(&self) -> bool {
+        self.semantic.source_type().is_typescript()
+    }
+
+    #[inline]
+    pub fn get_snippet(&self, span: Span) -> &'a str {
+        span.source_text(self.semantic.source_text())
+    }
+}
+
+impl<'a> PartialEq<IdentifierReference<'a>> for Symbol<'_, 'a> {
+    fn eq(&self, other: &IdentifierReference<'a>) -> bool {
+        // cheap: no resolved reference means its a global reference
+        let Some(reference_id) = other.reference_id.get() else {
+            return false;
+        };
+        // TODO: which is cheaper; a symbol_id check or a name check?
+        // if self.name() == other.name {
+        //     return true;
+        // }
```

Comment:
```suggestion
```

We must use `symbol_id` checking. The `IdentifierReference` of the same name in different scopes may not be equal.